### PR TITLE
Added "Step vector" radio button under "Display intervals in matrix as" (issue #617)

### PR DESCRIPTION
**Pull request by inthar-raven**

### CODE REVIEW

File: src/analysis.ts [hunk 1]
```diff
@@ -100,8 +100,9 @@ export function utonalFundamental(frequencies: number[], maxDivisor = 23) {
   return divisor * frequencies[0]
 }
 
-// Interval matrix a.k.a the modes of a scale
-export function intervalMatrix(scale: Scale) {
+// Interval matrix a.k.a the modes of a scale.
+export function intervalMatrix(scale: Scale): [[Interval]] {
```
Comment (frostburn) [inline src/analysis.ts:104]: This is not the correct type syntax.
image
Please use an editor with a working TypeScript plugin.

File: src/analysis.ts [hunk 2]
```diff
@@ -100,8 +100,9 @@ export function utonalFundamental(frequencies: number[], maxDivisor = 23) {
...
+  console.log('intervalMatrix was called')
```
Comment (frostburn) [inline src/analysis.ts:105]: No `console.log` in PRs without marking them as drafts.

File: src/analysis.ts [hunk 3]
```diff
@@ -110,3 +111,78 @@ export function intervalMatrix(scale: Scale) {
...
+  for (let i = 0; i < scale.size; ++i) {
+    const mode = scale.rotate(i)
+    result.push(columns.map((j) => getStepVector(mode, j)))
+  }
+  return result
+}
+
+// Get j-step vector on tonic of `scale` and return it as an array of [String, number] pairs
+// (sorted in descending order by size).
+export function getStepVector(scale: Scale, intervalClass: number): [[String, number]] {
```
Comment (frostburn) [inline src/analysis.ts:130]: Re-calculating the step vectors for each cell in the matrix is a waste of compute.

**Comment by frostburn** [comment]
My suggestion for the size indicators (smallest to largest).

1. x
2. s, L
3. s, m, L
4. s, m, M, L
5. s, m, n, M, L
6. s, S, m, M, l, L
7. s, S, m, n, M, l, L
8. s, S, m, n, N, M, l, L
9. a, b, c, d, e, f, g, h, i
10. (alphabet up to 10 first letters)
11. (etc.)

File: src/analysis.ts [hunk 4]
```diff
@@ -110,3 +111,78 @@ export function intervalMatrix(scale: Scale) {
...
+  const scaleAsSteps = [...Array(scale.intervals.length).keys()].map((i) =>
+    scale.rotate(i).getInterval(1)
+  ) // the scale as steps
+  const sortedUniqueSteps = scaleAsSteps.filter(function (item, pos) {
+    // This function asks, "Is it the first `pos` such that `rep(item)` is that value?"
+    // (I had to do this because of the whole Proxy vs. Object weirdness.)
+    return scaleAsSteps.findIndex((step) => step.toString() === item.toString()) == pos
+  })
+  // Sort the scale steps
+  sortedUniqueSteps.sort().reverse() // sort in descending order
```
Comment (frostburn) [inline src/analysis.ts:143]: `Array.sort` is a trap. You need to do `foo.sort((a, b) => a.compare(b))` or whatever I had implemented in `scale-workshop-core` if I had implemented anything...

File: src/analysis.ts [hunk 5]
```diff
@@ -110,3 +111,78 @@ export function intervalMatrix(scale: Scale) {
...
+      keys = ['L', 'm', 's']
+      break
+    }
+    case 4: {
+      keys = ['L', 'm', 'n', 's']
+      break
+    }
+    default: {
+      keys = ['L'].concat(
+        [...Array(sortedUniqueSteps.length - 2).keys()].map((i) => 'm'.concat(i + 1)),
```
Comment (frostburn) [inline src/analysis.ts:167]: Type error. Please don't abuse JS by concatenating integers to strings.

File: src/analysis.ts (hunk 1 above)
Comment (inthar-raven) [reply src/analysis.ts:104]: I think it's because I didn't import Interval. I should remove that declaration.

Comment (frostburn) [reply src/analysis.ts:104]: Then it's a valid type, but still the wrong one. The return type seems to be `Interval[][]`.

Comment (arseniiv) [reply src/analysis.ts:104]: That was syntax used in Haskell.

Comment (inthar-raven) [reply src/analysis.ts:104]: I just use Linux Mint's default text editor. The npm didn't catch any type errors, and I should have asked why.

Comment (frostburn) [reply src/analysis.ts:104]: `npm run type-check` does indeed pass. That's on Vue for writing a poor variant of TS.

File: src/analysis.ts (hunk 2 above)
Comment (inthar-raven) [reply src/analysis.ts:105]: Sorry, I forgot to check.

File: src/analysis.ts (hunk 3 above)
Comment (inthar-raven) [reply src/analysis.ts:130]: Yeah, I need to somehow do this "statically wrt the `Scale`".

**Comment by frostburn** [comment]
This PR proves the concept but it comes at a bad time:

- SW2 shouldn't be receiving new features that might cause new bugs or performance issues before EOL and the Big Freeze
- The code relies on `scale-workshop-core` and I'd need to do a lot of error-prone rebasing/porting to `sonic-weave` if this is merged before #619

I suggest adding the basic functionality for manipulating and formatting abstract periodic strings to https://github.com/xenharmonic-devs/moment-of-symmetry
We can then utilize it in SW3 when the beta cycle begins.

**Comment by inthar-raven** [comment]
> This PR proves the concept but it comes at a bad time:
> 
>     * SW2 shouldn't be receiving new features that might cause new bugs or performance issues before EOL and the Big Freeze
> 
>     * The code relies on `scale-workshop-core` and I'd need to do a lot of error-prone rebasing/porting to `sonic-weave` if this is merged before [Enter v3 beta cycle / The SW3 commit #619](https://github.com/xenharmonic-devs/scale-workshop/pull/619)
> 
> 
> I suggest adding the basic functionality for manipulating and formatting abstract periodic strings to https://github.com/xenharmonic-devs/moment-of-symmetry We can then utilize it in SW3 when the beta cycle begins.

I could first check if overhauling the core to `sonic-weave` actually breaks it.

**Comment by inthar-raven** [comment]
> I suggest adding the basic functionality for manipulating and formatting abstract periodic strings to https://github.com/xenharmonic-devs/moment-of-symmetry We can then utilize it in SW3 when the beta cycle begins.

Yeah. This new word manipulation stuff could use a new module, anyway.

**Comment by inthar-raven** [comment]
A couple more things we need to keep in mind when adapting this to SW3:
1. The variety for each interval class obviously needs to be at the step vector level in this mode
2. CS is never violated at the step vector level

**Comment by frostburn** [comment]
> A couple more things we need to keep in mind when adapting this to SW3:
> 
> 1. The variety for each interval class obviously needs to be at the step vector level in this mode
> 2. CS is never violated at the step vector level

1. I think we should show abstract variety vs. actual variety
2. Nuh-uh: all x violates CS. (And seriously we should show actual CS violation even if abstractly there is none.)

**Comment by inthar-raven** [comment]
>     2. Nuh-uh: all x violates CS. (And seriously we should show actual CS violation even if abstractly there is none.)
What I meant is that if a step vector has k steps, it obviously can never be another step vector with a different number of steps, though these two step vectors might be tuned the same. However, I must concede you have a point: CS violations at the tuning level would be useful to see.

**Pull request closed without merging**